demos/http3/ossl-nghttp3-demo-server.c

```diff
@@ -0,0 +1,723 @@
+    struct ssl_id *ssl_ids;
+    int i;
+
+    ssl_ids = h3ssl->ssl_ids;
+    for (i = 0; i < MAXSSL_IDS; i++) {
+        if (ssl_ids[i].s == NULL) {
+            ssl_ids[i].s = ssl;
+            ssl_ids[i].id = id;
+            return;
+        }
+    }
+    printf("Oops too many streams to add!!!\n");
+    exit(1);
+}
+
+static void h3close(struct h3ssl *h3ssl, uint64_t id) {
+    struct ssl_id *ssl_ids;
+    int i;
+
+    ssl_ids = h3ssl->ssl_ids;
+    for (i = 0; i < MAXSSL_IDS; i++) {
+        if (ssl_ids[i].id == id) {
+            if (!SSL_stream_conclude(ssl_ids[i].s, 0))
+                printf("h3close: SSL_stream_conclude failed\n");
+            SSL_shutdown(ssl_ids[i].s);
+        }
+    }
+}
+
+static int on_recv_header(nghttp3_conn *conn, int64_t stream_id, int32_t token,
+                          nghttp3_rcbuf *name, nghttp3_rcbuf *value,
+                          uint8_t flags, void *user_data,
+                          void *stream_user_data) {
+    nghttp3_vec vname, vvalue;
+
+    /* Received a single HTTP header. */
+    vname = nghttp3_rcbuf_get_buf(name);
+    vvalue = nghttp3_rcbuf_get_buf(value);
+
+    fwrite(vname.base, vname.len, 1, stderr);
+    fprintf(stderr, ": ");
+    fwrite(vvalue.base, vvalue.len, 1, stderr);
+    fprintf(stderr, "\n");
+
+    return 0;
+}
+
+static int on_end_headers(nghttp3_conn *conn, int64_t stream_id, int fin,
+                          void *user_data, void *stream_user_data) {
+    struct h3ssl *h3ssl = (struct h3ssl *)user_data;
+
+    fprintf(stderr, "on_end_headers!\n");
+    h3ssl->end_headers_received = 1;
+    return 0;
+}
+
+static int on_recv_data(nghttp3_conn *conn, int64_t stream_id,
+                        const uint8_t *data, size_t datalen,
+                        void *conn_user_data, void *stream_user_data) {
+    fprintf(stderr, "on_recv_data! %ld\n", (unsigned long)datalen);
+    fprintf(stderr, "on_recv_data! %.*s\n", (int)datalen, data);
+    return 0;
+}
+
+static int on_end_stream(nghttp3_conn *h3conn, int64_t stream_id,
+                         void *conn_user_data, void *stream_user_data) {
+    struct h3ssl *h3ssl = (struct h3ssl *)conn_user_data;
+
+    printf("on_end_stream!\n");
+    h3ssl->done = 1;
+    return 0;
+}
+
+static int read_from_ssl_ids(nghttp3_conn *h3conn, struct h3ssl *h3ssl) {
+    uint8_t msg2[16000];
+    int hassomething = 0, i;
+    struct ssl_id *ssl_ids;
+    SSL_POLL_ITEM items[MAXSSL_IDS] = {0}, *item = items;
+    static const struct timeval nz_timeout = {0, 0};
+    size_t result_count = SIZE_MAX;
+    int numitem = 0, ret;
+
+    item = items;
+    ssl_ids = h3ssl->ssl_ids;
+    /*
+     * Process all the streams
+     * the first one is the connection if we get something here is a new stream
+     */
+    for (i = 0; i < MAXSSL_IDS; i++) {
+        if (ssl_ids[i].s != NULL) {
+            item->desc = SSL_as_poll_descriptor(ssl_ids[i].s);
+            item->events = UINT64_MAX;
```

Comment:
Is this a way of setting all the bits?  If so, `~0` is perhaps more clear. Is there a more specific bitmask of events we might want here? A macro for these at the top of the file (even if defined as `~0`) may also be warranted.

Reply:
I copied that from ./test/quic_multistream_test.c.